File: source/python/expression_tree.py
class Node:
    def __init__(self, value, left=None, right=None):
        self.value = value
        self.left = left
        self.right = right

def evaluate(node):
    if node.left is None and node.right is None:
        return int(node.value)
    left_val = evaluate(node.left)
    right_val = evaluate(node.right)
    if node.value == "+":
        return left_val + right_val
    if node.value == "-":
        return left_val - right_val
    if node.value == "*":
        return left_val * right_val
    return 0

# Build: (2 + 3) * (4 - 1)
add_node = Node("+", Node("2"), Node("3"))
sub_node = Node("-", Node("4"), Node("1"))
mul_node = Node("*", add_node, sub_node)

result = evaluate(mul_node)
print(f"result: {result}")

File: source/javascript/expression_tree.js
// Expression tree evaluation
// Build tree for (2 + 3) * (4 - 1)
function newNode(value) { return { value, left: null, right: null }; }

function evaluate(node) {
    if (node.left === null && node.right === null) return parseInt(node.value);
    const leftVal = evaluate(node.left);
    const rightVal = evaluate(node.right);
    if (node.value === "+") return leftVal + rightVal;
    if (node.value === "-") return leftVal - rightVal;
    if (node.value === "*") return leftVal * rightVal;
    return 0;
}

const addNode = newNode("+");
addNode.left = newNode("2");
addNode.right = newNode("3");

const subNode = newNode("-");
subNode.left = newNode("4");
subNode.right = newNode("1");

const mulNode = newNode("*");
mulNode.left = addNode;
mulNode.right = subNode;

const result = evaluate(mulNode);
console.log(`result: ${result}`);
